sdk/iot/pnp/src/az_iot_pnp_client.c

```diff
@@ -53,32 +54,42 @@ AZ_NODISCARD az_result az_iot_pnp_client_init(
 
 AZ_NODISCARD az_result az_iot_pnp_client_get_user_name(
     az_iot_pnp_client const* client,
-    az_span mqtt_user_name,
-    az_span* out_mqtt_user_name)
+    char* mqtt_user_name,
+    size_t mqtt_user_name_size,
+    size_t* out_mqtt_user_name_length)
 {
   AZ_PRECONDITION_NOT_NULL(client);
-  AZ_PRECONDITION_VALID_SPAN(mqtt_user_name, 0, false);
-  AZ_PRECONDITION_NOT_NULL(out_mqtt_user_name);
+  AZ_PRECONDITION_NOT_NULL(mqtt_user_name);
+  AZ_PRECONDITION(mqtt_user_name_size > 0);
 
-  az_span remainder = mqtt_user_name;
-  az_span user_name;
-  AZ_RETURN_IF_FAILED(
-      az_iot_hub_client_user_name_get(&client->_internal.iot_hub_client, remainder, &user_name));
+  size_t written;
 
-  int32_t written = az_span_size(user_name);
-  remainder = az_span_slice_to_end(remainder, written);
+  // First get hub user name since it is unknown how long it will be
+  AZ_RETURN_IF_FAILED(az_iot_hub_client_user_name_get(
+      &client->_internal.iot_hub_client, mqtt_user_name, mqtt_user_name_size, &written));
 
-  int32_t required_length
-      = az_span_size(pnp_model_id) + az_span_size(client->_internal.root_interface_name) + 2;
+  az_span mqtt_user_name_span
+      = az_span_init((uint8_t*)mqtt_user_name, (int32_t)mqtt_user_name_size);
 
-  AZ_RETURN_IF_NOT_ENOUGH_SIZE(remainder, required_length);
+  int32_t required_length = (int32_t)written + (int32_t)sizeof(pnp_client_param_separator)
```

Comment:
We should consider whether this `required_size|length` should include the sizeof(null_terminalor) here.  I get we're trying to return what a strlen() would have returned on the output string buffer, but coming from Windows world at least they tend to include # of bytes and not length of string.  (See [this blast from the past for registry query](https://docs.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-regqueryvalueexa)].

If we want to just have the strlen() value of the string, OK I guess, but we need to document it on each and every place that we use this pattern.  I had to look this up to be sure almost every time I used a Windows (or preferabbly Windows CE :)) API.

Needs word-smithing but a pattern could be something like:
```
 * @param[in] mqtt_client_id_size The size, in bytes of \p mqtt_client_id.  Must include enough room for trailing \0.
 * @param[out] out_mqtt_client_id_length __[nullable]__ Contains the string length, in bytes, of
 *                                                      of \p mqtt_client_id, not counting trailing \0.  Can be `NULL`.
```

Reply:
Noted. For now we are not including the null terminator as `strlen()` does.